### ISSUE TITLE
Error:[One of the following attributes are set but missing the other for the density type: ldpi, mdpi, hdpi, xhdpi, xxhdpi, xxxhdpi. Please ensure that all require attributes are defined.]

### ISSUE THREAD

**Issue by Fmaldonado6** [apache/cordova-android#694]
I'm getting that error when I try to build my app eith an adaptive-icon in my config.xml, my code is the next one:

`<platform name="android">
    <resource-file src="res/icon/android/colors.xml" target="/app/src/main/res/values/colors.xml" />
    <icon background="@color/background" density="ldpi" foreground="res/icon/android/ldpi-foreground.png" />
    <icon background="@color/background" density="mdpi" foreground="res/icon/android/mdpi-foreground.png" />
    <icon background="@color/background" density="hdpi" foreground="res/icon/android/hdpi-foreground.png" />
    <icon background="@color/background" density="xhdpi" foreground="res/icon/android/xhdpi-foreground.png" />
    <icon background="@color/background" density="xxhdpi" foreground="res/icon/android/xxhdpi-foreground.png" />
    <icon background="@color/background" density="xxxhdpi" foreground="res/icon/android/xxxhdpi-foreground.png" />
</platform>`

I'm using cordova-android 8.0.0

**Comment by adamduren** [comment]
@Fmaldonado6 After some digging through the code I found that it was because I was using global cordova `cordova@8.1.2`, upgrading to `cordova@9.0.0` via `npm i -g cordova@9.0.0` fixed the problem for me. This is because even though we upgraded to `cordova-android@8.0.0` in our project this library depends on information passed to it from the globally installed `cordova` library which it gets from `cordova-common@3.1.0` whereas `cordova@8.1.2` has `cordova-common@2.2.5` which does not pass along `foreground` and `background` props

This was not indicated in the upgrade instructions AFAIK. It would be helpful as a fix to this bug for `cordova-android` to release a patch that warns if the features you need are not available in the global `cordova` that is installed.

Hope this helps.

**Comment by breautek** [comment]
Closing, solution posted above.

TL;DR; version: Ensure your cordova tooling is up to date.

**Comment by RayyanNafees** [comment]
AM working with cordova 9 still got the ssue
`<?xml version='1.0' encoding='utf-8'?>

<widget id="com.jaininfosoft.app" version="1.0.0" xmlns="http://www.w3.org/ns/widgets" xmlns:cdv="http://cordova.apache.org/ns/1.0">
    <name>JainInfoSoft</name>
    <description>
        A sample Apache Cordova application that responds to the deviceready event.
    </description>
    <author email="jaininfosoft.com" href="http://cordova.io">
        Jain Info Soft
    </author>
    <content src="index.html" />
    <access origin="*" />
    <allow-intent href="http://*/*" />
    <allow-intent href="https://*/*" />
    <allow-intent href="tel:*" />
    <allow-intent href="sms:*" />
    <allow-intent href="mailto:*" />
    <allow-intent href="geo:*" />
    
    <platform name="android">
        <allow-intent href="market:*" />
        <icon src="./www/assets/logo.svg" background="white" width="57" height="57" density="mdpi" type="mdpi" platform="android"/>
        <splash density="mdpi" src="./www/assets/imgs/splash.svg" width="360" />
    </platform>

    <platform name="ios">
        <allow-intent href="itms:*" />
        <allow-intent href="itms-apps:*" />
    </platform>
</widget>
`

Error while `cordova prepare`:
One of the following attributes are set but missing the other for the density type: mdpi. Please ensure that all require attributes are defined.

**Comment by faugusztin** [comment]
What happens if you remove the type="mdpi" attribute from your icon definition ? Icon has no such attribute according to [documentation](https://cordova.apache.org/docs/en/10.x/config_ref/images.html#page-toc-source). Also you are supposed to use the foreground attribute too if you are using the background attribute for adaptive icons. In general, your icon definition doesn't match the documentation (this didn't change between Cordova 9 & 10).
